kani-compiler/src/codegen_cprover_gotoc/codegen/rvalue.rs

```diff
@@ -747,31 +753,108 @@ impl<'tcx> GotocCtx<'tcx> {
+                if src_mir_type == dst_mir_type {
+                    src_goto_expr
+                } else {
+                    self.codegen_unsized_cast(src_goto_expr, src_mir_type, dst_mir_type)
+                }
             }
         }
     }
 
-    fn cast_to_unsized_expr(
+    /// Generate code for unsized cast. This includes the following:
+    /// -> (Built-in / Smart) Pointer from array to slice.
+    /// -> (Built-in / Smart) Pointer from sized type to dyn trait.
+    /// -> (Built-in / Smart) Pointer from dyn trait to dyn trait.
+    ///     - E.g.: `(dyn Any + Send)` to `&dyn Any`.
+    /// -> All the cases above where the pointer refers to a parametrized struct where the type
+    /// parameter is the target of the unsize casting.
+    ///     - E.g.: `RcBox<String>` to `RcBox<dyn Any>`
+    fn codegen_unsized_cast(
         &mut self,
         src_goto_expr: Expr,
         src_mir_type: Ty<'tcx>,
         dst_mir_type: Ty<'tcx>,
-    ) -> Option<Expr> {
-        // Check if the cast is from a vtable fat pointer to another
-        // vtable fat pointer (which can happen with auto trait fat pointers)
-        if self.is_vtable_fat_pointer(src_mir_type) {
-            self.cast_unsized_dyn_trait_to_unsized_dyn_trait(
-                src_goto_expr,
-                src_mir_type,
-                dst_mir_type,
-            )
-        } else {
-            // Recursively cast the source expression into an unsized expression.
-            // This will include thin pointers, slices, and Adt.
-            self.cast_expr_to_unsized_expr(src_goto_expr, src_mir_type, dst_mir_type)
-        }
+    ) -> Expr {
+        // Collect some information about the unsized coercion.
+        let mut path = self.collect_unsized_cast_path(src_goto_expr, src_mir_type, dst_mir_type);
+        debug!(cast=?path, "codegen_unsized_cast");
+
+        // Handle the leaf which should always be a pointer.
+        let (ptr_cast_info, ptr_src_expr) = path.pop().unwrap();
+        let initial_expr = self.codegen_cast_to_fat_pointer(ptr_src_expr, ptr_cast_info);
+
+        // Iterate from the back of the path initializing each struct that requires the coercion.
+        // This code is required for handling smart pointers.
+        path.into_iter().rev().fold(initial_expr, |coercion_expr, (info, src_expr)| {
+            self.codegen_struct_unsized_coercion(src_expr, info, coercion_expr)
+        })
+    }
+
+    /// Extract path that must be explicitly casted. Add to the tuple the expression type for
+    /// the current source.
+    fn collect_unsized_cast_path(
+        &self,
+        src_goto_expr: Expr,
+        src_mir_type: Ty<'tcx>,
+        dst_mir_type: Ty<'tcx>,
+    ) -> Vec<(CoerceUnsizedInfo<'tcx>, Expr)> {
+        let mut field_type = src_goto_expr;
+        CoerceUnsizedIterator::new(self.tcx, src_mir_type, dst_mir_type)
+            .map(|info| {
+                let expr = if let Some(field_symbol) = info.field {
```

Comment:
What does it mean if info doesn't have a field?

Reply:
It means we are no longer traversing a structure. I.e.: we found the underlying pointer.

Reply:
I added some comments to see if it helps.